Lib/asyncio/locks.py

```diff
@@ -403,7 +402,22 @@ def release(self):
         become larger than zero again, wake up that coroutine.
         """
         self._value += 1
-        self._wake_up_next()
+        self._wake_up_first()
+
+    def _wake_up_first(self):
+        """Wake up the first waiter if it isn't done."""
+        if not self._waiters:
+            return
+        try:
+            fut = next(iter(self._waiters))
+        except StopIteration:
+            return
+
+        # .done() necessarily means that a waiter will wake up later on and
+        # either take the lock, or, if it was cancelled and lock wasn't
+        # taken already, will hit this again and wake up a new waiter.
+        if not fut.done():
+            fut.set_result(True)
```

Comment:
Unlike Lock, in Semaphore you may call release() repeatedly. If there are several waiters, some of which have already been woken up (but not removed their Future yet), this should wake up the next one. This actually makes the code simpler (you may want to keep some of the comments, paraphrased):
```suggestion
        for fut in self._waiters:
            if not fut.done():
                fut.set_result(True)
                break
```
This removes the need for at least one sleep(0.01) in my testing, maybe more.

Reply:
The name of `_wake_up_first` implies it only wakes up the _first_ waiter. If we wake up multiple waiters at once, how do we know they will resume from `await` in FIFO order?

Reply:
After some trials I think what actually happens is worse than my previous comment. This change would result in negative counter in certain cases and this is a bug. An example is when you use `0.01` in test file Line 850. To explain it, the two `release()` would each awake a waiter. Then the first waiter awakes a third waiter because the second waiter hasn't yet decreased the counter. But there really isn't a value for the third waiter.

The test can pass when using time `0` simply because the execution hasn't reached far enough into the bug. In other words, the test does not completely test the entire execution. This probably suggests using `sleep(0.01)` to test is better in some circumstance. Otherwise we'd have to use multiple `sleep(0)`.

Reply:
Yeah, I think I see the problem with my suggestion now. We *either* wake up the next waiter that isn't done yet in `release()`, and then `acquire()` only needs to wake someone when it is cancelled; *or* we do it your way. Both are correct. The difference is that your way requires two event loop iterations to wake two waiters, even in the test case that does `release(); release()`. The other way requires only one event loop iteration but relies on the event loop running tasks in order. So let's do it your way, but use multiple `sleep(0)` calls to make the tests pass instead of `sleep(0.01)` -- at least in the new tests that you've added, and in those tests that had a single `sleep(0)` before. Existing tests that were using `sleep(0.01)` can be fixed in a separate PR.